net/raft/raft_test.go

```diff
@@ -116,3 +120,119 @@ func TestStartUninitialized(t *testing.T) {
 		t.Errorf("sv.Exec() = %s, want %s", err, ErrUninitialized)
 	}
 }
+
+func TestClusterSetup(t *testing.T) {
+	ctx := context.Background()
+
+	// Create three uninitialized raft services.
+	nodeA := newTestNode(t)
+	defer nodeA.cleanup()
+	nodeB := newTestNode(t)
+	defer nodeB.cleanup()
+	nodeC := newTestNode(t)
+	defer nodeC.cleanup()
+
+	// Initialize A, creating a fresh cluster.
+	must(t, nodeA.service.Init())
+
+	// Update the cluster to allow A, B and C's addresses.
+	var err error
+	_, err = nodeA.service.Exec(ctx, set("/allowed/"+nodeA.addr, "yes"))
+	must(t, err)
+	_, err = nodeA.service.Exec(ctx, set("/allowed/"+nodeB.addr, "yes"))
+	must(t, err)
+	_, err = nodeA.service.Exec(ctx, set("/allowed/"+nodeC.addr, "yes"))
+	must(t, err)
+
+	// Add B and C to the cluster.
+	must(t, nodeB.service.Join("https://"+nodeA.addr))
+	must(t, nodeC.service.Join("https://"+nodeA.addr))
+
+	// Try setting a value on nodeB.
+	_, err = nodeB.service.Exec(ctx, set("/foo", "bar"))
+	must(t, err)
+
+	// Try reading the value on nodeC's state.
+	must(t, nodeC.service.WaitRead(ctx))
+	got := nodeC.state.Data["/foo"]
+	if got != "bar" {
+		t.Errorf("reading /foo, nodeC got %q want %q", got, "bar")
+	}
+}
+
+func must(t *testing.T, err error) {
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+type testNode struct {
+	dir    string
+	addr   string
+	server *httptest.Server
+	state  *state
+
+	wg      sync.WaitGroup
+	service *Service
+}
+
+func (n *testNode) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	n.wg.Wait()
+	n.service.ServeHTTP(rw, req)
+}
+
+func (n *testNode) cleanup() {
+	os.RemoveAll(n.dir)
```

Comment:
I don't think there's anything we can do here if cleaning up the test fails. It's also okay to be best effort because newTestNode uses a temporary directory which the OS will eventually cleanup.